File: src/PalindromePartition.java
import java.util.ArrayList;
import java.util.List;
//Time Complexity : O(2^N)
//Space Complexity : O(N) 
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : No

/**
 * Apply backtracking. Start iterating from 0th index of the string. fetch the
 * substring from start to end (which starts from start to length of the string)
 * and check if it is palindrome. If so, add it to the list of palindrome
 * strings and then call Backtracking from next index. To cover all
 * possibilities, we have to drop the last added palidrom from the list. Add all
 * these lists when the end of the string is reached.
 *
 */
class Solution {
	List<List<String>> ans = new ArrayList<>();

	public List<List<String>> partition(String s) {
		helper(s, 0, new ArrayList<>());
		return ans;
	}

	public void helper(String s, int start, List<String> list) {
		if (start >= s.length()) {
			ans.add(new ArrayList<>(list));
			return;
		}

		for (int end = start; end < s.length(); end++) {
			if (isPalindrome(s, start, end)) {
				list.add(s.substring(start, end + 1));
				helper(s, end + 1, list);
				list.remove(list.size() - 1);
			}
		}
	}

	public boolean isPalindrome(String s, int i, int j) {
		while (i <= j) {
			if (s.charAt(i) != s.charAt(j))
				return false;
			i++;
			j--;
		}
		return true;
	}
}

File: Palindrome_Partition.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    /// Time Complexity : O(N * 2^N) 
    // Space Complexity :O(N) 
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  No 
    public class Palindrome_Partition
    {
        IList<IList<string>> result;
        public IList<IList<string>> Partition(string s)
        {
            result = new List<IList<string>>();
            if (s == null || s.Length == 0) return result;
            helper(s, 0, new List<string>());
            return result;
        }

        private void helper(string s, int idx, List<string> path)
        {
            //base
            if (idx == s.Length)
            {
                result.Add(new List<string>(path));
                return;
            }
            //logic

            for (int i = idx; i < s.Length; i++)
            {
                int length = i - idx;
                string sub = s.Substring(idx, length + 1);
                if (IsPalindrome(sub))
                {
                    //action
                    path.Add(sub);
                    //recurse
                    helper(s, i + 1, path);
                    //back track
                    path.RemoveAt(path.Count - 1);
                }
            }
        }

        private bool IsPalindrome(string s)
        {
            int start = 0;
            int end = s.Length - 1;
            while (start < end)
            {
                if (s.ToCharArray()[start] != s.ToCharArray()[end])
                    return false;
                start++;
                end--;
            }

            return true;
        }
    }
}
